pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+		case buildapi.BuildPhaseComplete:
+			bc.recorder.Eventf(patchedBuild, kapi.EventTypeNormal, buildapi.BuildCompletedEventReason, fmt.Sprintf(buildapi.BuildCompletedEventMessage, patchedBuild.Namespace, patchedBuild.Name))
+		case buildapi.BuildPhaseError,
+			buildapi.BuildPhaseFailed:
+			bc.recorder.Eventf(patchedBuild, kapi.EventTypeNormal, buildapi.BuildFailedEventReason, fmt.Sprintf(buildapi.BuildFailedEventMessage, patchedBuild.Namespace, patchedBuild.Name))
+		}
+		if isTerminal(patchedBuild.Status.Phase) {
+			common.HandleBuildCompletion(patchedBuild, bc.buildLister, bc.buildConfigGetter, bc.buildDeleter, bc.runPolicies)
+		}
+	}
+}
+
+// patchBuild generates a patch for the given build and buildUpdate
+// and applies that patch using the REST client
+func (bc *BuildController) patchBuild(build *buildapi.Build, update *buildUpdate) (*buildapi.Build, error) {
+
+	// Create a patch using the buildUpdate object
+	updatedBuild, err := buildutil.BuildDeepCopy(build)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create a deep copy of build %s: %v", buildDesc(build), err)
+	}
+	update.apply(updatedBuild)
+	patch, err := validation.CreateBuildPatch(build, updatedBuild)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a build patch: %v", err)
+	}
+
+	glog.V(5).Infof("Patching build %s with %v", buildDesc(build), update)
+	return bc.buildPatcher.Patch(build.Namespace, build.Name, patch)
+}
+
+// findMissingPod uses the REST client directly to determine if a pod exists or not.
+// It is called when a corresponding pod for a build is not found in the cache.
+func (bc *BuildController) findMissingPod(build *buildapi.Build) *kapi.Pod {
+	// Make one last attempt to fetch the pod using the REST client
+	pod, err := bc.podClient.Pods(build.Namespace).Get(buildapi.GetBuildPodName(build), metav1.GetOptions{})
+	if err == nil {
+		glog.V(2).Infof("Found missing pod for build %s by using direct client.", buildDesc(build))
+		return pod
+	}
+	return nil
+}
+
+// getPodByKey looks up a pod by key in the podInformer cache
+func (bc *BuildController) getPodByKey(key string) (*kapi.Pod, error) {
+	obj, exists, err := bc.podInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		glog.V(2).Infof("Unable to retrieve pod %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return nil, err
+	}
+	if !exists {
+		glog.V(2).Infof("Pod %q has been deleted", key)
+		return nil, nil
+	}
+
+	return obj.(*kapi.Pod), nil
+}
+
+// getBuildByKey looks up a build by key in the buildInformer cache
+func (bc *BuildController) getBuildByKey(key string) (*buildapi.Build, error) {
+	obj, exists, err := bc.buildInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		glog.V(2).Infof("Unable to retrieve build %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
```

Comment:
maybe remove this, and call bc.handleError() in work() if bc.getBuildByKey() fails?